text/0048-synchronizer-redesign.md

```diff
@@ -0,0 +1,170 @@
+* downloading dependencies as fast as possible to enable quick finalization of blocks.
+
+The current implementation started off as a naive way of doing this, and has received some isolated improvements, but still makes redundant requests sometimes and is not as efficient as it could be.
+
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The synchronizer keeps collections of pending vertices that cannot be added to the protocol state yet, because:
+
+* their timestamp is from the future,
+* they are missing a dependency, or
+* they are waiting for validation of the block they contain.
+
+Based on these collections, and the peers we received them from, we can deduce which missing dependencies we need to request, and whom we expect to have them: a correct node will only send us a vertex if it already has all dependencies.
+
+To keep the synchronizer's memory usage small and its computations fast, we want to keep its state as small as possible, so we want to prioritize vertices in a way that minimizes the time we expect them to remain pending in the synchronizer: We always want to request those vertices first which are most likely to be added to the protocol state soon. That roughly means: first all evidence against faulty validators, since evidence never has a dependency and can always be added to the protocol state right away; then all endorsements and units, starting with those with an early timestamp, since they tend to have fewer dependencies.
+
+Based on that prioritization we want to send requests so that:
+
+* We request each vertex from only one peer at a time, but with an aggressive timeout.
+* We don't request too many vertices from the same peer at once. The number will be configurable. A low default — even 1 or 2 — should be fine for a running node, but we should determine experimentally whether a higher value speeds up syncing for a joining node.
+
+Finally, adding vertices to the protocol state as soon as possible is preferable over keeping them in the synchronizer queues: It allows full verification and earlier finality detection. So when syncing, we prefer to download the protocol state bottom-up: the earlier vertices before the later ones. That is why we will add a new way to specify a dependency, so that if we receive unit number 200 from a peer, and we only have unit 100 from that creator yet, we can request unit 101 instead of having to start from unit 199 and moving backwards.
+
+
+## Reference-level explanation
+
+We will move the synchronizer into the `highway_core` module, so it has access to the full content of each vertex, and all internal Highway data.
+
+
+### New dependency type: "Unit with sequence number"
+
+In addition to specifying units by hash, we add a message type
+
+```rust
+UnitBySeqNumber(u64, Hash)
+```
+
+which means we ask for a unit with the specified sequence number by the same sender as the unit with the given hash.
```

Comment:
What if the sender gives us a unit with the correct sequence number but a different parent? i.e. a unit from a different fork. This will make the node start downloading units from a different chain, back to the beginning of an era.

Reply:
In the first step of the second queue processing pass it says we use this only if "the unit's creator is not known to be faulty". So what will happen if Alice created `a0, a1, a2, a3, a4`, but also a fork `a0', a1', a2', a3'`, and we have `a0, a1`, and she sends us `a4'`? We'll request "unit number 2", since our latest one is `a1`. If she sends us `a2'`, that will be earlier in the queue, since it has an earlier timestamp, so we'll _first_ try to request `a2'`'s dependencies. That includes `a1'`, which we'd request _by hash_. But as soon as we have `a1'` and `a1`, we know that Alice is faulty, so we will stop requesting units by sequence number from her.